<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scRNAseq analysis functions. Developed using Seurat framework. â€¢ scMiko</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="scRNAseq analysis functions. Developed using Seurat framework.">
<meta property="og:description" content="Set of utility functions used to conduct scRNAseq analysis. Build on functionality of Seurat Package. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scMiko</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Computational Resources</li>
    <li>
      <a href="articles/install.html">scMiko Package</a>
    </li>
    <li>
      <a href="articles/scPipeline_getting_started.html">scPipeline Reports</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">scMiko Vignettes</li>
    <li>
      <a href="articles/DEG.html">Differential-expression using CDI</a>
    </li>
    <li>
      <a href="articles/Cluster_Optimization.html">Selecting cluster resolution using specificity criterion</a>
    </li>
    <li>
      <a href="articles/Marker_Catalog.html">Cell-type marker catalog</a>
    </li>
    <li>
      <a href="articles/Cell_Annotation.html">Marker-based cell-type annotation using Miko Scoring</a>
    </li>
    <li>
      <a href="articles/Module_Detection.html">Gene program discovery using SSN analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">scPipeline Dashboard Reports</li>
    <li>
      <a href="articles/scPipeline_module_overview.html">Overview of modules</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/NMikolajewicz/scMiko">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level2">
<h2 id="about-scmiko">About scMiko<a class="anchor" aria-label="anchor" href="#about-scmiko"></a>
</h2>
<p><a href="https://github.com/NMikolajewicz/scMiko">scMiko</a> is an R package developed for single-cell RNA-seq analysis. It was designed using the <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a> framework, and offers existing and novel single-cell analytic work flows.</p>
<p>Refer to our <a href="articles/install.html">getting started with scMiko</a> article for installation instructions.</p>
<p>New features offered in scMiko include:</p>
<ul>
<li>Differential expression analysis using the <strong>co-dependency index (CDI)</strong>
<ul>
<li>CDI identified binary DEGs</li>
<li>binary DEGs are highly-specific</li>
<li><a href="articles/DEG.html">Vignette</a></li>
</ul>
</li>
<li>Specificity-based cluster <strong>resolution selection criterion</strong>
<ul>
<li>method to identify optimal clustering configurations</li>
<li>out-performs resampling-based approaches</li>
<li><a href="articles/Cluster_Optimization.html">Vignette</a></li>
</ul>
</li>
<li>Cell-type annotation using <strong>Miko Scoring pipeline</strong>
<ul>
<li>unbiased scoring of variably-sized gene sets</li>
<li>hypothesis-testing frame work that accepts or rejects of candidate annotations</li>
<li><a href="articles/Cell_Annotation.html">Vignette</a></li>
</ul>
</li>
<li>Gene program discovery using <strong>scale-free shared-nearest neighbor network (SSN) analysis</strong>
<ul>
<li>gene program discovery algorithm to identify and functionally annotate gene programs in an unsupervised manner</li>
<li>outperforms ICA and NMF methods in terms of GO term recovery and STRING PPI enrichment</li>
<li><a href="articles/Module_Detection.html">Vignette</a></li>
</ul>
</li>
</ul>
<p>Details can be found in our recent work: <a href="https://www.biorxiv.org/content/10.1101/2022.03.13.484162v1" class="external-link">Mikolajewicz, et al., BioRxiv</a></p>
</div>
<div class="section level2">
<h2 id="about-scpipeline">About scPipeline<a class="anchor" aria-label="anchor" href="#about-scpipeline"></a>
</h2>
<p>The <a href="https://github.com/NMikolajewicz/scPipeline">scPipeline</a> analysis pipeline is a modular collection of R markdown scripts that generate <a href="https://rstudio.github.io/flexdashboard/index.html" class="external-link">flexdashboard</a> reports for scRNAseq analyses, and is dependent on the <a href="https://github.com/NMikolajewicz/scMiko">scMiko</a> package. The modular framework permits flexible usage, without necessitating complete reliance on a single analysis platform. Additionally, the self-contained reports that are generated by each module provide a comprehensive analysis summary and log of analytic parameters and scripts, thereby ensuring reproducible and shareable analysis workflows.</p>
<p>Refer to our <a href="articles/scPipeline_getting_started.html">getting started with scPipeline</a> article for installation and configuration instructions, and to our <a href="articles/scPipeline_module_overview.html">scPipeline overview</a> article for detailed module overviews and examples of analyses.</p>
<p>Currently the following analysis modules are available:</p>
<ul>
<li>
<a href="articles/scPipeline_module_overview.html#module-1-quality-control-qc-preprocessing-1">Module 1</a>: QC &amp; preprocessing</li>
<li>
<a href="articles/scPipeline_module_overview.html#module-2-data-integration-1">Module 2</a>: Integration</li>
<li>
<a href="articles/scPipeline_module_overview.html#module-3-cluster-optimization-1">Module 3</a>: Cluster optimization</li>
<li>
<a href="articles/scPipeline_module_overview.html#module-4-cell-type-annotation-1">Module 4</a>: Cell annotation</li>
<li>
<a href="articles/scPipeline_module_overview.html#module-5-gene-program-discovery-1">Module 5</a>: Gene program discovery</li>
<li>
<a href="articles/scPipeline_module_overview.html#module-6-gene-expression-and-associations-1">Module 6</a>: Gene expression and association</li>
</ul>
<p><a href="https://github.com/NMikolajewicz/scMiko"><strong>scMiko</strong></a> and <a href="https://github.com/NMikolajewicz/scPipeline"><strong>scPipeline</strong></a> are developed and maintained by the Moffat lab.</p>
</div>
<div class="section level2">
<h2 id="authors">Authors<a class="anchor" aria-label="anchor" href="#authors"></a>
</h2>
<ul>
<li><a href="https://scholar.google.ca/citations?user=LBWQMXsAAAAJ&amp;hl=en&amp;oi=ao" class="external-link">Nicholas Mikolajewicz</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<ul>
<li>Hong Han and Kevin Brown from the <a href="http://moffatlab.ccbr.utoronto.ca/" class="external-link">Moffat Lab</a>
</li>
</ul>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scMiko</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://scholar.google.ca/scholar?hl=en&amp;as_sdt=0%2C5&amp;q=mikolajewicz+nicholas&amp;oq=" class="external-link">Nicholas Mikolajewicz</a> <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://scholar.google.ca/scholar?hl=en&amp;as_sdt=0%2C5&amp;q=mikolajewicz+nicholas&amp;oq=" class="external-link">Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
