<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Identifying Differentially Expressed Genes using the Co-Dependence Index • scMiko</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Identifying Differentially Expressed Genes using the Co-Dependence Index"><meta property="og:description" content="scMiko"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMiko</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Computational Resources</li>
    <li>
      <a href="../articles/install.html">scMiko Package</a>
    </li>
    <li>
      <a href="../articles/scPipeline_getting_started.html">scPipeline Reports</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">scMiko Vignettes</li>
    <li>
      <a href="../articles/DEG.html">Differential-expression using CDI</a>
    </li>
    <li>
      <a href="../articles/Cluster_Optimization.html">Selecting cluster resolution using specificity criterion</a>
    </li>
    <li>
      <a href="../articles/Marker_Catalog.html">Cell-type marker catalog</a>
    </li>
    <li>
      <a href="../articles/Cell_Annotation.html">Marker-based cell-type annotation using Miko Scoring</a>
    </li>
    <li>
      <a href="../articles/Module_Detection.html">Gene program discovery using SSN analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">scPipeline Dashboard Reports</li>
    <li>
      <a href="../articles/scPipeline_module_overview.html">Overview of modules</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/NMikolajewicz/scMiko">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>





<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Identifying Differentially Expressed Genes using
the Co-Dependence Index</h1>
            
            <h4 data-toc-skip class="date">Compiled: 2022-08-03</h4>
      
      
      <div class="hidden name"><code>DEG.Rmd</code></div>

    </div>

    
    
<div id="load-seurat-object" class="section level2">
<h2>Load Seurat object</h2>
<p>For this tutorial we will analyze the Human Gastrulation dataset
reported by Tyser 2021. The dataset consists of 1195 cells. We will
perform differential expression analysis using two methods:
co-dependency index (CDI) and wilcoxon test.</p>
<p>We start by reading in the data and visualizing the annotated
population. We are interested in identifying cell-type specific
markers.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load package</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scMiko)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># load human gastrulation data</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>so.query <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;../data/demo/so_tyser2021_220621.rds&quot;</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize clusters</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cluster.UMAP</span>(<span class="at">so =</span> so.query, <span class="at">group.by =</span> <span class="st">&quot;sub_cluster&quot;</span>) <span class="sc">+</span> <span class="fu">theme_void</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Tyser 2021&quot;</span>,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">&quot;Human Gastrulation&quot;</span>)</span></code></pre></div>
<p><img src="C:\Users\Owner\Dropbox\PDFPRO~2\RPACKA~1\scMiko\docs\articles\DEG_FI~1/figure-html/load data-1.png" width="672" /></p>
</div>
<div id="perform-differential-gene-expression-analysis"
class="section level2">
<h2>Perform differential gene expression analysis</h2>
<p>To identifying cell-type specific markers, we will perform CDI- and
Wilcoxon-based differentially-expressed gene (DEG) analyses. Since CDI
is more computationally intensive, we will focus on the subset of genes
that are expressed within the dataset.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get expressed genes</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>expr_gene <span class="ot">&lt;-</span> <span class="fu">getExpressedGenes</span>(<span class="at">object =</span> so.query)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># CDI</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>cdi_output <span class="ot">&lt;-</span> <span class="fu">findCDIMarkers</span>(<span class="at">object =</span> so.query, <span class="at">features.x =</span> <span class="st">&quot;sub_cluster&quot;</span>, <span class="at">features.y =</span> expr_gene)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Wilcoxon</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>wilcoxon_output <span class="ot">&lt;-</span> <span class="fu">getDEG</span>(<span class="at">object =</span> so.query, <span class="at">group_by =</span> <span class="st">&quot;sub_cluster&quot;</span>, <span class="at">auc.thresh =</span> <span class="cn">NA</span>, <span class="at">fdr.thresh =</span> <span class="cn">NA</span>,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">return.all =</span> T, <span class="at">return.list =</span> F)</span></code></pre></div>
<p>Once the analysis has be complete, we can visualize the top
differentially-expressed genes using Seurat’s <code>DotPlot</code>
function.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get top markers</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>wilcoxon_output.top <span class="ot">&lt;-</span> wilcoxon_output <span class="sc">%&gt;%</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(group) <span class="sc">%&gt;%</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">1</span>, <span class="at">wt =</span> auc)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>cdi_output<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">gsub</span>(<span class="st">&quot;sub_cluster_&quot;</span>, <span class="st">&quot;&quot;</span>, cdi_output<span class="sc">$</span>feature.x), <span class="at">levels =</span> <span class="fu">unique</span>(wilcoxon_output.top<span class="sc">$</span>group))</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>cdi_output <span class="ot">&lt;-</span> cdi_output <span class="sc">%&gt;%</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(group)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>cdi_output.top <span class="ot">&lt;-</span> cdi_output <span class="sc">%&gt;%</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(feature.x) <span class="sc">%&gt;%</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">1</span>, <span class="at">wt =</span> ncdi)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="co"># generate dot plots</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>plt.cdi_dot <span class="ot">&lt;-</span> <span class="fu">DotPlot</span>(<span class="at">object =</span> so.query, <span class="at">features =</span> <span class="fu">unique</span>(cdi_output.top<span class="sc">$</span>feature.y), <span class="at">group.by =</span> <span class="st">&quot;sub_cluster&quot;</span>) <span class="sc">+</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_miko</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;CDI Markers&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Clusters&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Genes&quot;</span>) <span class="sc">+</span> <span class="fu">theme_miko</span>(<span class="at">legend =</span> T,</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">x.axis.rotation =</span> <span class="dv">45</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;bottom&quot;</span>)</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>plt.wilcoxon_dot <span class="ot">&lt;-</span> <span class="fu">DotPlot</span>(<span class="at">object =</span> so.query, <span class="at">features =</span> <span class="fu">unique</span>(wilcoxon_output.top<span class="sc">$</span>feature), <span class="at">group.by =</span> <span class="st">&quot;sub_cluster&quot;</span>) <span class="sc">+</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_miko</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Wilcoxon Markers&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Clusters&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Genes&quot;</span>) <span class="sc">+</span> <span class="fu">theme_miko</span>(<span class="at">legend =</span> T,</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">x.axis.rotation =</span> <span class="dv">45</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;bottom&quot;</span>)</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(plt.cdi_dot, plt.wilcoxon_dot, <span class="at">nrow =</span> <span class="dv">1</span>))</span></code></pre></div>
<p><img src="C:\Users\Owner\Dropbox\PDFPRO~2\RPACKA~1\scMiko\docs\articles\DEG_FI~1/figure-html/visualize-1.png" width="1056" /></p>
<p>Finally, we can compare the relative sensitivity and specificity of
markers obtained by each method.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get top 10 markers per cluster</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>cdi_output.top <span class="ot">&lt;-</span> cdi_output <span class="sc">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(feature.x) <span class="sc">%&gt;%</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">wt =</span> ncdi)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>wilcoxon_output.top <span class="ot">&lt;-</span> wilcoxon_output <span class="sc">%&gt;%</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(group) <span class="sc">%&gt;%</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">wt =</span> auc)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="co"># get sensitivity and specificity indices that were calculated by getDEG()</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>cdi_sensitivity <span class="ot">&lt;-</span> wilcoxon_output <span class="sc">%&gt;%</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="fu">paste0</span>(feature, <span class="st">&quot;_&quot;</span>, group) <span class="sc">%in%</span> <span class="fu">paste0</span>(cdi_output.top<span class="sc">$</span>feature.y, <span class="st">&quot;_&quot;</span>, cdi_output.top<span class="sc">$</span>group))</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>cdi_sensitivity<span class="sc">$</span>method <span class="ot">&lt;-</span> <span class="st">&quot;CDI&quot;</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>wilcoxon_sensitivity <span class="ot">&lt;-</span> wilcoxon_output <span class="sc">%&gt;%</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="fu">paste0</span>(feature, <span class="st">&quot;_&quot;</span>, group) <span class="sc">%in%</span> <span class="fu">paste0</span>(wilcoxon_output.top<span class="sc">$</span>feature, <span class="st">&quot;_&quot;</span>, wilcoxon_output.top<span class="sc">$</span>group))</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>wilcoxon_sensitivity<span class="sc">$</span>method <span class="ot">&lt;-</span> <span class="st">&quot;Wilcoxon&quot;</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>df.eval <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(cdi_sensitivity, wilcoxon_sensitivity)</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="co"># generate plots</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>plt_sensitivity <span class="ot">&lt;-</span> df.eval <span class="sc">%&gt;%</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> method, <span class="at">y =</span> sensitivity, <span class="at">fill =</span> method)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>() <span class="sc">+</span> ggbeeswarm<span class="sc">::</span><span class="fu">geom_quasirandom</span>() <span class="sc">+</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_miko</span>(<span class="at">fill.palette =</span> <span class="st">&quot;ptol&quot;</span>) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Sensitivity&quot;</span>) <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>))</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>plt_specificity <span class="ot">&lt;-</span> df.eval <span class="sc">%&gt;%</span></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> method, <span class="at">y =</span> specificity, <span class="at">fill =</span> method)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>() <span class="sc">+</span> ggbeeswarm<span class="sc">::</span><span class="fu">geom_quasirandom</span>() <span class="sc">+</span></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_miko</span>(<span class="at">fill.palette =</span> <span class="st">&quot;ptol&quot;</span>) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Specificity&quot;</span>) <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>))</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(plt_sensitivity, plt_specificity)</span></code></pre></div>
<p><img src="C:\Users\Owner\Dropbox\PDFPRO~2\RPACKA~1\scMiko\docs\articles\DEG_FI~1/figure-html/compare sensitivity and specificity-1.png" width="768" /></p>
</div>
<div id="adjusting-for-batch-effects" class="section level2">
<h2>Adjusting for batch effects</h2>
<p>In many instances, single-cell data from multiple independent
biological replicates are pooled. To minimize batch effects and ensure
the validity of the CDI across independent samples, we recommend one of
two alternative approaches:</p>
<ul>
<li>Depth-normalization (across samples)</li>
<li>Fisher’s method (meta-analysis)</li>
</ul>
<p>We will use the Pijuan-Sala (2019) gastrulation data to demonstrate
each approach. First we load in the count matrix, split it into
component datasets and normalize each, and then merge the normalized
data into a single Seurat object. Splitting the Seurat object into
component data sets allows us to fit a SCT Model to each dataset.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load murine gastrulation data</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>ps.data <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;../data/demo/ps_count_matrix.rds&quot;</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>ps.meta <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;../data/demo/ps_meta_data.rds&quot;</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># create seurat object</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>so.ps <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> ps.data, <span class="at">meta.data =</span> ps.meta)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># split by sample</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>so.ps.list <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(<span class="at">object =</span> so.ps, <span class="at">split.by =</span> <span class="st">&quot;sample&quot;</span>)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="co"># sample-wise normalization of data</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>so.ps.list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="at">X =</span> so.ps.list, <span class="at">FUN =</span> SCTransform, <span class="at">method =</span> <span class="st">&quot;glmGamPoi&quot;</span>, <span class="at">verbose =</span> F, <span class="at">vst.flavor =</span> <span class="st">&quot;v2&quot;</span>,</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">variable.features.rv.th =</span> <span class="fl">1.3</span>)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="co"># merge normalized data</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>so.ps <span class="ot">&lt;-</span> <span class="fu">merge</span>(so.ps.list[[<span class="dv">1</span>]], so.ps.list[<span class="sc">-</span><span class="dv">1</span>])</span></code></pre></div>
<div id="approach-1-depth-normalization" class="section level3">
<h3>Approach 1: Depth-normalization</h3>
<p>In the first approach, we will use the Seurat object with multiple
independent samples and preprocess it using Seurat’s
PrepSCTFindMarkers() function which effectively downsamples the count
matrix to ensure a homogeneous sequencing depth across all component
datasets. The processed Seurat data will then be provided as input into
findCDIMarkers():</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># enforce homogeneous sequencing depth across all samples</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>so.ps_homogenous <span class="ot">&lt;-</span> <span class="fu">PrepSCTFindMarkers</span>(so.ps)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># run CDI on heterogeneous sequencing depth data</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>cdi.heterogenous <span class="ot">&lt;-</span> <span class="fu">findCDIMarkers</span>(so.ps, <span class="at">features.x =</span> <span class="st">&quot;celltype&quot;</span>, <span class="at">verbose =</span> F)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co"># run CDI on homogeneous sequencing depth data</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>cdi.homogeneous <span class="ot">&lt;-</span> <span class="fu">findCDIMarkers</span>(so.ps_homogenous, <span class="at">features.x =</span> <span class="st">&quot;celltype&quot;</span>, <span class="at">verbose =</span> F)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="co"># merge CDI results from both runs for comparison</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>cdi.merge.approach1 <span class="ot">&lt;-</span> <span class="fu">merge</span>(cdi.heterogenous, cdi.homogeneous, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;feature.x&quot;</span>, <span class="st">&quot;feature.y&quot;</span>))</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="co"># plot CDI scores from depth-normalized run to visulaize how much results have changed</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>r.val <span class="ot">&lt;-</span> <span class="fu">signif</span>(<span class="fu">cor</span>(cdi.merge.approach1<span class="sc">$</span>ncdi.x, cdi.merge.approach1<span class="sc">$</span>ncdi.y), <span class="dv">3</span>)</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>plt.comparison.approach1 <span class="ot">&lt;-</span> cdi.merge.approach1 <span class="sc">%&gt;%</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> ncdi.x, <span class="at">y =</span> ncdi.y)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Batch correction using Approach 1&quot;</span>,</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;nCDI (using non-corrected counts)&quot;</span>, <span class="at">y =</span> <span class="st">&quot;nCDI (using corrected counts)&quot;</span>, <span class="at">subtitle =</span> <span class="fu">paste0</span>(<span class="st">&quot;Pearson&#39;s r = &quot;</span>,</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>        r.val)) <span class="sc">+</span> <span class="fu">theme_miko</span>(<span class="at">legend =</span> T) <span class="sc">+</span> <span class="fu">geom_abline</span>(<span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="at">color =</span> <span class="st">&quot;tomato&quot;</span>)</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>plt.comparison.approach1</span></code></pre></div>
<p><img src="C:\Users\Owner\Dropbox\PDFPRO~2\RPACKA~1\scMiko\docs\articles\DEG_FI~1/figure-html/batch correction approahc 1-1.png" width="700" /></p>
</div>
<div id="approach-2-fishers-method" class="section level3">
<h3>Approach 2: Fisher’s Method</h3>
<p>In the second approach, we will apply a meta-analytic approach to
pool CDI results from independent samples thereby finding markers that
are conserved across samples. P values from independent sample-specific
estimates are pooled using the Fisher’s Method.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run CDI for each sample and pool the sample-level statistics</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>cdi.conserved <span class="ot">&lt;-</span> <span class="fu">findConservedCDIMarkers</span>(<span class="at">object =</span> so.ps, <span class="at">features.x =</span> <span class="st">&quot;celltype&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;sample&quot;</span>,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.workers =</span> <span class="dv">15</span>, <span class="at">verbose =</span> F)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># merge CDI results from both runs for comparison</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>cdi.merge.approach2 <span class="ot">&lt;-</span> <span class="fu">merge</span>(cdi.heterogenous, cdi.conserved, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;feature.x&quot;</span>, <span class="st">&quot;feature.y&quot;</span>))</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># plot CDI scores from conserved CDI run to visulaize how much results have changed</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>r.val.approach2 <span class="ot">&lt;-</span> <span class="fu">signif</span>(<span class="fu">cor</span>(cdi.merge.approach2<span class="sc">$</span>ncdi, cdi.merge.approach2<span class="sc">$</span>ncdi_mean), <span class="dv">3</span>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>plt.comparison.approach2 <span class="ot">&lt;-</span> cdi.merge.approach2 <span class="sc">%&gt;%</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> ncdi, <span class="at">y =</span> ncdi_mean)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Batch correction using Approach 2&quot;</span>,</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;nCDI (using non-corrected counts)&quot;</span>, <span class="at">y =</span> <span class="st">&quot;nCDI (averaged across samples)&quot;</span>, <span class="at">subtitle =</span> <span class="fu">paste0</span>(<span class="st">&quot;Pearson&#39;s r = &quot;</span>,</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>        r.val.approach2)) <span class="sc">+</span> <span class="fu">theme_miko</span>(<span class="at">legend =</span> T) <span class="sc">+</span> <span class="fu">geom_abline</span>(<span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="at">color =</span> <span class="st">&quot;tomato&quot;</span>)</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>plt.comparison.approach2</span></code></pre></div>
<p><img src="C:\Users\Owner\Dropbox\PDFPRO~2\RPACKA~1\scMiko\docs\articles\DEG_FI~1/figure-html/batch correction approahc 2-1.png" width="700" /></p>
</div>
<div id="comparison-of-batch-normalized-cdi-results"
class="section level3">
<h3>Comparison of batch-normalized CDI results</h3>
<p>Here we compare the batch-normalized CDI results (approach 1 and 2)
with uncorrected CDI results. Using a 5% FDR threshold, the overlap in
significant gene sets is found to vary depending on which approach is
taken. Here we show that Approach 1 yielded more consistent results with
the uncorrected CDI results, whereas Approach 2 yielded less consistent
results. In the latter, the results suggest the presence of
batch-specific artefacts that may have contributed to slightly different
results in the uncorrected CDI analysis. Although Approach 2 is more
computationally intensive, it is rooted in established meta-analytic
theory and should be employed when evaluated independent experimental
samples.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get DEGs at 5% FDR</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>DEG.uncorrected <span class="ot">&lt;-</span> <span class="fu">longDF2namedList</span>(cdi.heterogenous <span class="sc">%&gt;%</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(fdr <span class="sc">&lt;</span> <span class="fl">0.05</span>), <span class="at">group_by =</span> <span class="st">&quot;feature.x&quot;</span>, <span class="at">values =</span> <span class="st">&quot;feature.y&quot;</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>DEG.approach1 <span class="ot">&lt;-</span> <span class="fu">longDF2namedList</span>(cdi.homogeneous <span class="sc">%&gt;%</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(fdr <span class="sc">&lt;</span> <span class="fl">0.05</span>), <span class="at">group_by =</span> <span class="st">&quot;feature.x&quot;</span>, <span class="at">values =</span> <span class="st">&quot;feature.y&quot;</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>DEG.approach2 <span class="ot">&lt;-</span> <span class="fu">longDF2namedList</span>(cdi.conserved <span class="sc">%&gt;%</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(fdr <span class="sc">&lt;</span> <span class="fl">0.05</span>), <span class="at">group_by =</span> <span class="st">&quot;feature.x&quot;</span>, <span class="at">values =</span> <span class="st">&quot;feature.y&quot;</span>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># annotate method used</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(DEG.uncorrected) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;uncorrected_&quot;</span>, <span class="fu">names</span>(DEG.uncorrected))</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(DEG.approach1) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;approach1_&quot;</span>, <span class="fu">names</span>(DEG.approach1))</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(DEG.approach2) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;approach2_&quot;</span>, <span class="fu">names</span>(DEG.approach2))</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="co"># combine DEG lists and compute jaccard similarities</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>DEG.list <span class="ot">&lt;-</span> <span class="fu">c</span>(DEG.uncorrected, DEG.approach1, DEG.approach2)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>jmat <span class="ot">&lt;-</span> <span class="fu">jaccardSimilarityMatrix</span>(DEG.list)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>unc_a1 <span class="ot">&lt;-</span> jmat[<span class="fu">grepl</span>(<span class="st">&quot;uncorrected&quot;</span>, <span class="fu">rownames</span>(jmat)), <span class="fu">grepl</span>(<span class="st">&quot;approach1&quot;</span>, <span class="fu">rownames</span>(jmat))]</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>unc_a2 <span class="ot">&lt;-</span> jmat[<span class="fu">grepl</span>(<span class="st">&quot;uncorrected&quot;</span>, <span class="fu">rownames</span>(jmat)), <span class="fu">grepl</span>(<span class="st">&quot;approach2&quot;</span>, <span class="fu">rownames</span>(jmat))]</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>a1_a2 <span class="ot">&lt;-</span> jmat[<span class="fu">grepl</span>(<span class="st">&quot;approach1&quot;</span>, <span class="fu">rownames</span>(jmat)), <span class="fu">grepl</span>(<span class="st">&quot;approach2&quot;</span>, <span class="fu">rownames</span>(jmat))]</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>unc_a1.max <span class="ot">&lt;-</span> <span class="fu">apply</span>(unc_a1, <span class="dv">1</span>, max)</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>unc_a2.max <span class="ot">&lt;-</span> <span class="fu">apply</span>(unc_a2, <span class="dv">1</span>, max)</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>a1_a2.max <span class="ot">&lt;-</span> <span class="fu">apply</span>(a1_a2, <span class="dv">1</span>, max)</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="co"># combine jaccard similarity profiles from each method comparison</span></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>df.deg.overlap <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(<span class="fu">data.frame</span>(<span class="at">Group_1 =</span> <span class="st">&quot;uncorrected&quot;</span>, <span class="at">Group_2 =</span> <span class="st">&quot;approach1&quot;</span>, <span class="at">DEG_overlap =</span> unc_a1.max),</span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(<span class="at">Group_1 =</span> <span class="st">&quot;uncorrected&quot;</span>, <span class="at">Group_2 =</span> <span class="st">&quot;approach2&quot;</span>, <span class="at">DEG_overlap =</span> unc_a2.max), <span class="fu">data.frame</span>(<span class="at">Group_1 =</span> <span class="st">&quot;approach1&quot;</span>,</span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>        <span class="at">Group_2 =</span> <span class="st">&quot;approach2&quot;</span>, <span class="at">DEG_overlap =</span> a1_a2.max))</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a><span class="co"># plot jaccard similarity profiles between batch corrected CDI methods</span></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>plt.deg.overlap <span class="ot">&lt;-</span> df.deg.overlap <span class="sc">%&gt;%</span></span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">paste0</span>(Group_1, <span class="st">&quot;_vs_&quot;</span>, Group_2), <span class="at">y =</span> DEG_overlap)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>(<span class="at">fill =</span> <span class="st">&quot;grey&quot;</span>) <span class="sc">+</span></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>    ggbeeswarm<span class="sc">::</span><span class="fu">geom_quasirandom</span>() <span class="sc">+</span> <span class="fu">theme_miko</span>() <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;DEG Overlap&quot;</span>, <span class="at">subtitle =</span> <span class="st">&quot;Jaccard Similarity of DEGs (per cluster)&quot;</span>,</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Jaccard Similarity&quot;</span>, <span class="at">x =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>plt.deg.overlap</span></code></pre></div>
<p><img src="C:\Users\Owner\Dropbox\PDFPRO~2\RPACKA~1\scMiko\docs\articles\DEG_FI~1/figure-html/compare batch-normalized CDI results-1.png" width="700" /></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
      </div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://scholar.google.ca/scholar?hl=en&amp;as_sdt=0%2C5&amp;q=mikolajewicz+nicholas&amp;oq=" class="external-link">Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  


  

  </body></html>
