<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Gene expression markers for all identity classes. — FindAllMarkers.Parallel • scMiko</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Gene expression markers for all identity classes. — FindAllMarkers.Parallel" />
<meta property="og:description" content="Finds markers (differentially expressed genes) for each of the identity classes in a dataset, with parallel implementation. Produces identical output as Seurat::FindAllMarkers," />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMiko</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Gene expression markers for all identity classes.</h1>
    
    <div class="hidden name"><code>FindAllMarkers.Parallel.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Finds markers (differentially expressed genes) for each of the identity classes in a dataset, with parallel implementation. Produces identical output as Seurat::FindAllMarkers,</p>
    </div>

    <pre class="usage"><span class='fu'>FindAllMarkers.Parallel</span><span class='op'>(</span>
  <span class='va'>object</span>,
  n.work <span class='op'>=</span> <span class='fl'>1</span>,
  assay <span class='op'>=</span> <span class='fu'>DefaultAssay</span><span class='op'>(</span><span class='va'>object</span><span class='op'>)</span>,
  slot <span class='op'>=</span> <span class='st'>"data"</span>,
  only.pos <span class='op'>=</span> <span class='cn'>F</span>,
  min.pct <span class='op'>=</span> <span class='fl'>0</span>,
  test.use <span class='op'>=</span> <span class='st'>"MAST"</span>,
  logfc.threshold <span class='op'>=</span> <span class='fl'>0.15</span>,
  max.cells.per.ident <span class='op'>=</span> <span class='fl'>200</span>,
  return.thresh <span class='op'>=</span> <span class='fl'>1</span>,
  verbose <span class='op'>=</span> <span class='cn'>F</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Seurat object</p></td>
    </tr>
    <tr>
      <th>n.work</th>
      <td><p>Number of workers. If 1 (default), runs Seurat::FindAllMarkers(). If &gt;1, parallelized implementation is used.</p></td>
    </tr>
    <tr>
      <th>assay</th>
      <td><p>Assay to use in differential expression testing. Default is DefaultAssay(object)</p></td>
    </tr>
    <tr>
      <th>slot</th>
      <td><p>Slot to pull data from; note that if test.use is "negbinom", "poisson", or "DESeq2", slot will be set to "counts". Default is "data".</p></td>
    </tr>
    <tr>
      <th>only.pos</th>
      <td><p>Only return positive markers (FALSE by default)</p></td>
    </tr>
    <tr>
      <th>min.pct</th>
      <td><p>only test genes that are detected in a minimum fraction of min.pct cells in either of the two populations. Meant to speed up the function by not testing genes that are very infrequently expressed. Default is 0</p></td>
    </tr>
    <tr>
      <th>test.use</th>
      <td><p>Test to use. See Seurat::FindAllMarkers() or Seurat::FindMarkers() for list of all options.</p></td>
    </tr>
    <tr>
      <th>logfc.threshold</th>
      <td><p>Limit testing to genes which show, on average, at least X-fold difference (log-scale) between the two groups of cells. Default is 0.25 Increasing logfc.threshold speeds up the function, but can miss weaker signals.</p></td>
    </tr>
    <tr>
      <th>max.cells.per.ident</th>
      <td><p>Down sample each identity class to a max number. Default is 200.</p></td>
    </tr>
    <tr>
      <th>return.thresh</th>
      <td><p>Only return markers that have a p-value &lt; return.thresh, or a power &gt; return.thresh (if the test is ROC).</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Print a progress bar once expression testing begins. Default is F.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments passed to Seurat::FindAllMarkers() (if n.work = 1) or Seurat::FindMarkers() (if n.work &gt;1)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>FindAllMarkers</code> <code>FindMarkers</code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Nicholas Mikolajewicz</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'># Find markers for all clusters</span>
<span class='va'>all.markers</span> <span class='op'>&lt;-</span> <span class='fu'>FindAllMarkers.Parallel</span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>so.query</span>, n.work <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocParallel</span></div><div class='output co'>#&gt; <span class='error'>Error in unique(object@active.ident): object 'so.query' not found</span></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://scholar.google.ca/scholar?hl=en&as_sdt=0%2C5&q=mikolajewicz+nicholas&oq='>Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


