<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference • scMiko</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMiko</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/install.html">Getting Started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">scMiko Vignettes</li>
    <li>
      <a href="../articles/DEG.html">Differentially-Expressed Genes</a>
    </li>
    <li>
      <a href="../articles/Cluster_Optimization.html">Cluster Resolution Optimization</a>
    </li>
    <li>
      <a href="../articles/Marker_Catalog.html">Cell-Type Marker Catalog</a>
    </li>
    <li>
      <a href="../articles/Cell_Annotation.html">Cell-Type Annotation</a>
    </li>
    <li>
      <a href="../articles/Module_Detection.html">Gene Module Detection</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">scPipeline Dashboard Reports</li>
    <li>
      <a href="../articles/R643_M01_Ochoka2021_QC_NM2_21_Sep_3_39_09_PM_2021.html">QC and Preprocessing Module</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/NMikolajewicz/scMiko">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-filters-and-qc" class="hasAnchor"><a href="#section-filters-and-qc" class="anchor"></a>Filters and QC</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="filterSeurat.html">filterSeurat()</a></code> </p>
        </td>
        <td><p>Apply QC filters to Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clusterFilter.html">clusterFilter()</a></code> </p>
        </td>
        <td><p>Filter seurat object by specified cluster ids</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cleanCluster.html">cleanCluster()</a></code> </p>
        </td>
        <td><p>Clean clusters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cleanFilterGenes.html">cleanFilterGenes()</a></code> </p>
        </td>
        <td><p>Clean and filter gene list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findArtifactGenes.html">findArtifactGenes()</a></code> </p>
        </td>
        <td><p>Identify artifact genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getExpressedGenes.html">getExpressedGenes()</a></code> </p>
        </td>
        <td><p>Identify expressed genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getExpressingCells.html">getExpressingCells()</a></code> </p>
        </td>
        <td><p>Get cells that express query gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMitoContent.html">getMitoContent()</a></code> </p>
        </td>
        <td><p>Calculate mitochondrial content</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="QC.scatterPlot.html">QC.scatterPlot()</a></code> </p>
        </td>
        <td><p>QC scatter plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="QC.violinPlot.html">QC.violinPlot()</a></code> </p>
        </td>
        <td><p>QC violin plots</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-dimensional-reduction" class="hasAnchor"><a href="#section-dimensional-reduction" class="anchor"></a>Dimensional Reduction</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="pcaElbow.html">pcaElbow()</a></code> </p>
        </td>
        <td><p>Quickly estimate the 'elbow' of a scree plot (PCA)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="projectReduction.html">projectReduction()</a></code> </p>
        </td>
        <td><p>Project dimensionally-reduced features onto UMAP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="propVarPCA.html">propVarPCA()</a></code> </p>
        </td>
        <td><p>Variance explained by each principal component.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getReductionGenes.html">getReductionGenes()</a></code> </p>
        </td>
        <td><p>Get top loaded features for PCA or ICA dimensional reduction.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runRPCA.html">runRPCA()</a></code> </p>
        </td>
        <td><p>Run Robust Prinicipal Component Analysis</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-clustering" class="hasAnchor"><a href="#section-clustering" class="anchor"></a>Clustering</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="setResolution.html">setResolution()</a></code> </p>
        </td>
        <td><p>Set cluster resolution</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multiCluster.html">multiCluster()</a></code> </p>
        </td>
        <td><p>Cluster seurat object at several resolutions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multiSpecificity.html">multiSpecificity()</a></code> </p>
        </td>
        <td><p>Evaluate specificity of single-cell markers across several cluster resolutions.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multiSilhouette.html">multiSilhouette()</a></code> </p>
        </td>
        <td><p>Evaluate silhouette indices of clustered single cell data across several cluster resolutions.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-cell-type-annotation" class="hasAnchor"><a href="#section-cell-type-annotation" class="anchor"></a>Cell-Type Annotation</h2>
          <p class="section-desc"><p>Module detection</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="mikoScore.html">mikoScore()</a></code> </p>
        </td>
        <td><p>Calculate Miko module scores for feature expression programs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="nullScore.html">nullScore()</a></code> </p>
        </td>
        <td><p>Null distribution for standardized module scores.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sigScore.html">sigScore()</a></code> </p>
        </td>
        <td><p>Calculate Miko score significance</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-gene-set-functions" class="hasAnchor"><a href="#section-gene-set-functions" class="anchor"></a>Gene Set Functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="AddSModuleScore.html">AddSModuleScore()</a></code> </p>
        </td>
        <td><p>Calculate standardized module scores for feature expression programs in single cells.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aggGroupExpression.html">aggGroupExpression()</a></code> </p>
        </td>
        <td><p>Get summary of group expression in Seurat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="avgGroupExpression.html">avgGroupExpression()</a></code> </p>
        </td>
        <td><p>Get summary of group expression in Seurat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="coherentFraction.html">coherentFraction()</a></code> </p>
        </td>
        <td><p>Calculate coherent fraction for feature expression program.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="consolidateNMF.html">consolidateNMF()</a></code> </p>
        </td>
        <td><p>Consolidate several NMF reduction objects into single NMF reduction object within Seurat object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getJaccard.html">getJaccard()</a></code> </p>
        </td>
        <td><p>Jaccard Similarity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="id2geneset.html">id2geneset()</a></code> </p>
        </td>
        <td><p>Get Reactome/GO geneset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="id2term.html">id2term()</a></code> </p>
        </td>
        <td><p>Map Reactome/GO ID to term</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="jaccardSimilarityMatrix.html">jaccardSimilarityMatrix()</a></code> </p>
        </td>
        <td><p>Jaccard Similarity Matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="optimalBinSize.html">optimalBinSize()</a></code> </p>
        </td>
        <td><p>Identify optimal bin size for AddModuleScore() function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runAUC.html">runAUC()</a></code> </p>
        </td>
        <td><p>Run AUCell classification</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runMS.html">runMS()</a></code> </p>
        </td>
        <td><p>Run Modular Scoring</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarizeHG.html">summarizeHG()</a></code> </p>
        </td>
        <td><p>Summarize hypergeometric enrichment results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="baderPathways.html">`{`()</a></code> </p>
        </td>
        <td><p>Pathway annotations from Bader Lab</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAnnotationPathways.html">getAnnotationPathways()</a></code> </p>
        </td>
        <td><p>Returns list of annotations for given Entrez gene IDs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runGSEA.html">runGSEA()</a></code> </p>
        </td>
        <td><p>Run gene-set enrichment analysis (GSEA)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runHG.html">runHG()</a></code> </p>
        </td>
        <td><p>Run hypergeometric gene enrichment analysis.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="searchAnnotations.html">searchAnnotations()</a></code> </p>
        </td>
        <td><p>Search Reactome/GO databases for terms that match query</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="signatureCoherence.html">signatureCoherence()</a></code> </p>
        </td>
        <td><p>Evaluate signature coherence.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="term2id.html">term2id()</a></code> </p>
        </td>
        <td><p>Map Reactome/GO term to ID</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="upset.Plot.html">upset.Plot()</a></code> </p>
        </td>
        <td><p>Upset plot</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-expression-functions" class="hasAnchor"><a href="#section-expression-functions" class="anchor"></a>Expression Functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="findCDIMarkers.html">findCDIMarkers()</a></code> </p>
        </td>
        <td><p>Calculate feature dependency index</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findGiniMarkers.html">findGiniMarkers()</a></code> </p>
        </td>
        <td><p>Calculate Gini marker specificity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDEG.html">getDEG()</a></code> </p>
        </td>
        <td><p>Get differentially expressed genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getExpressionMatrix.html">getExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Get expression matrix from Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="miko_volcano.html">miko_volcano()</a></code> </p>
        </td>
        <td><p>Draw volcano plot to visualize differential expression.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parCor.html">parCor()</a></code> </p>
        </td>
        <td><p>Parallelized correlation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pseudotimeRF.html">pseudotimeRF()</a></code> </p>
        </td>
        <td><p>Identify pseudotime-dependent genes using Random Forest (RF) Model</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-variance-decomposition" class="hasAnchor"><a href="#section-variance-decomposition" class="anchor"></a>Variance Decomposition</h2>
          <p class="section-desc"><p>Variance decomposition analyses</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="vd_Formula.html">vd_Formula()</a></code> </p>
        </td>
        <td><p>Specify model formula for variance decomposition.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vd_Inputs.html">vd_Inputs()</a></code> </p>
        </td>
        <td><p>Specify inputs for variance decomposition analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vd_Run.html">vd_Run()</a></code> </p>
        </td>
        <td><p>Perform Variance Decomposition Analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vd_UMAP.html">vd_UMAP()</a></code> </p>
        </td>
        <td><p>Generate UMAPs with each variance decomposition covariate overlaid.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-scale-free-shared-nearest-neighbor-network-analysis-ssn-" class="hasAnchor"><a href="#section-scale-free-shared-nearest-neighbor-network-analysis-ssn-" class="anchor"></a>Scale-Free Shared Nearest Neighbor Network Analysis (SSN)</h2>
          <p class="section-desc"><p>Module detection</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="SSNConnectivity.html">SSNConnectivity()</a></code> </p>
        </td>
        <td><p>SSN connectivity plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SSNResolution.html">SSNResolution()</a></code> </p>
        </td>
        <td><p>Identify optimal cluster resolution of scale-free shared nearest neighbor network (SNN)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pruneSSN.html">pruneSSN()</a></code> </p>
        </td>
        <td><p>Identify and (optionally) prune module features in scale-free shared nearest neighbor network (SSN)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runSSN.html">runSSN()</a></code> </p>
        </td>
        <td><p>Run SNN</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SSNExpression.html">SSNExpression()</a></code> </p>
        </td>
        <td><p>SSN connectivity plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findNetworkFeatures.html">findNetworkFeatures()</a></code> </p>
        </td>
        <td><p>Identify gene module features.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getModuleGenes.html">getModuleGenes()</a></code> </p>
        </td>
        <td><p>Get list of module genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaleFreeNet.html">scaleFreeNet()</a></code> </p>
        </td>
        <td><p>Apply scale-free topology transform to shared-nearest neighbor graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarizeModules.html">summarizeModules()</a></code> </p>
        </td>
        <td><p>Summarize SSN, ICA or NMF gene module analyses</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-independent-component-analysis" class="hasAnchor"><a href="#section-independent-component-analysis" class="anchor"></a>Independent Component Analysis</h2>
          <p class="section-desc"><p>ICA Module detection</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="getICAGenes.html">getICAGenes()</a></code> </p>
        </td>
        <td><p>Get significant ICA genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runICA.html">runICA()</a></code> </p>
        </td>
        <td><p>Run Independent Component Analysis on gene expression</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-non-negative-matrix-factorization" class="hasAnchor"><a href="#section-non-negative-matrix-factorization" class="anchor"></a>Non-Negative Matrix Factorization</h2>
          <p class="section-desc"><p>NMF Module detection</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="getNMFGenes.html">getNMFGenes()</a></code> </p>
        </td>
        <td><p>Returns top module genes from NMF feature loading matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runNMF.html">runNMF()</a></code> </p>
        </td>
        <td><p>Perform non-negative matrix factorization (NMF)</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-wgcna" class="hasAnchor"><a href="#section-wgcna" class="anchor"></a>WGCNA</h2>
          <p class="section-desc"><p>WGCNA Module detection</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="getModulePreservation.html">getModulePreservation()</a></code> </p>
        </td>
        <td><p>Determine module preservation between reference and query network</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSoftThreshold.html">getSoftThreshold()</a></code> </p>
        </td>
        <td><p>Analysis of scale free topology for soft-threshold</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSoftThreshold2.html">getSoftThreshold2()</a></code> </p>
        </td>
        <td><p>Analysis of scale free topology for soft-threshold. Modified from getSoftThreshold.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="optimalDS.html">optimalDS()</a></code> </p>
        </td>
        <td><p>Cut clustered tree at varying heights and overlay with dendrogram to find optimal parameter set.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runWGCNA.html">runWGCNA()</a></code> </p>
        </td>
        <td><p>run WGCNA analysis on scRNAseq expression matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaleTOM.html">scaleTOM()</a></code> </p>
        </td>
        <td><p>TOM matrix rescaling prior to computing consensus topological overlap.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wgcna2graphDF.html">wgcna2graphDF()</a></code> </p>
        </td>
        <td><p>Convert adjacency or TOM matrix to igraph data.frame</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-differential-abundance-analysis" class="hasAnchor"><a href="#section-differential-abundance-analysis" class="anchor"></a>Differential Abundance Analysis</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="da_DEG.html">da_DEG()</a></code> </p>
        </td>
        <td><p>Identify correlated genes/pathways associated with differential abundance.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="da_Run.html">da_Run()</a></code> </p>
        </td>
        <td><p>Run differential abundance analysis.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-visualization" class="hasAnchor"><a href="#section-visualization" class="anchor"></a>Visualization</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="cluster.UMAP.html">cluster.UMAP()</a></code> </p>
        </td>
        <td><p>UMAP stratified by cluster ID</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="expression.Plot.html">expression.Plot()</a></code> </p>
        </td>
        <td><p>Violin plot of single cell gene expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="exprUMAP.html">exprUMAP()</a></code> </p>
        </td>
        <td><p>Visualize gene expression on UMAP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="featureGradient.html">featureGradient()</a></code> </p>
        </td>
        <td><p>Visualize feature activity/expression gradient overlaid on UMAP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneRepCurve.html">geneRepCurve()</a></code> </p>
        </td>
        <td><p>Plot relationship showing percentage of cells expressing atleast percentage of genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geom_split_violin.html">geom_split_violin()</a></code> </p>
        </td>
        <td><p>Split violin plot using ggplot2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getUMAP.html">getUMAP()</a></code> </p>
        </td>
        <td><p>Get UMAP data and plot from Seurat object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="miko_heatmap.html">miko_heatmap()</a></code> </p>
        </td>
        <td><p>Function to draw ggplot heatmaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scale_color_miko.html">scale_color_miko()</a></code> </p>
        </td>
        <td><p>Gradient color scale</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scale_fill_miko.html">scale_fill_miko()</a></code> </p>
        </td>
        <td><p>Gradient fill scale</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scExpression.UMAP.html">scExpression.UMAP()</a></code> </p>
        </td>
        <td><p>Cell-level gene expression projected on UMAP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="theme_miko.html">theme_miko()</a></code> </p>
        </td>
        <td><p>scMiko Theme</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="variableGenes.Plot.html">variableGenes.Plot()</a></code> </p>
        </td>
        <td><p>Plot variable genes in Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="highlightUMAP.html">highlightUMAP()</a></code> </p>
        </td>
        <td><p>Highlight cells on UMAP plot</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data" class="hasAnchor"><a href="#section-data" class="anchor"></a>Data</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="geneSets.html">geneSets</a></code> </p>
        </td>
        <td><p>List of scMiko gene sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="LR.db.html">LR.db</a></code> </p>
        </td>
        <td><p>Ligand-Receptor Database</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-miscellaneous-utilities" class="hasAnchor"><a href="#section-miscellaneous-utilities" class="anchor"></a>Miscellaneous Utilities</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="autoPointSize.html">autoPointSize()</a></code> </p>
        </td>
        <td><p>Automatically determine optimal point size for geom_point()</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="balanceMatrixSize.html">balanceMatrixSize()</a></code> </p>
        </td>
        <td><p>Balance matrix dimensions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="categoricalColPal.html">categoricalColPal()</a></code> </p>
        </td>
        <td><p>Generate categorical ColorBrewer palette.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="citationCheck.html">citationCheck()</a></code> </p>
        </td>
        <td><p>Check pubmed citations for genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clearGlobalEnv.html">clearGlobalEnv()</a></code> </p>
        </td>
        <td><p>Remove variables from global environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="col2rowname.html">col2rowname()</a></code> </p>
        </td>
        <td><p>Assign column entries in data.frame to row names.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dist2hclust.html">dist2hclust()</a></code> </p>
        </td>
        <td><p>Hierarchially-cluster distance matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fixBarcodeLabel.html">fixBarcodeLabel()</a></code> </p>
        </td>
        <td><p>Fix barcode labels</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getClusterCenters.html">getClusterCenters()</a></code> </p>
        </td>
        <td><p>Get cluster centers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getConnectivity.html">getConnectivity()</a></code> </p>
        </td>
        <td><p>Gene connectivity within network.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDensity.html">getDensity()</a></code> </p>
        </td>
        <td><p>Get local density (z) of bivariate relationship (x,y)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getNodesEdges.html">getNodesEdges()</a></code> </p>
        </td>
        <td><p>Get nodes and edges from igraph data.frame for visNetwork</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getOrderedGroups.html">getOrderedGroups()</a></code> </p>
        </td>
        <td><p>Get vector of unique ordered group names from Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="group2list.html">group2list()</a></code> </p>
        </td>
        <td><p>Named list of cells grouped by meta feature from Seurat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lintersect.html">lintersect()</a></code> </p>
        </td>
        <td><p>Returns intersection of all list entries.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="longDF2namedList.html">longDF2namedList()</a></code> </p>
        </td>
        <td><p>Convert long data frame to named list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="miko_message.html">miko_message()</a></code> </p>
        </td>
        <td><p>Print message</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="namedList2longDF.html">namedList2longDF()</a></code> </p>
        </td>
        <td><p>Convert named list to long data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="namedList2wideDF.html">namedList2wideDF()</a></code> </p>
        </td>
        <td><p>Convert named list to wide data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="orderedFactor.html">orderedFactor()</a></code> </p>
        </td>
        <td><p>Sort factor levels in numerical order</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pseudoReplicates.html">pseudoReplicates()</a></code> </p>
        </td>
        <td><p>Create pseudo-replicates, stratified by grouping variable.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qNorm.html">qNorm()</a></code> </p>
        </td>
        <td><p>Quantile Normalization of 2 Vectors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rescaleValues.html">rescaleValues()</a></code> </p>
        </td>
        <td><p>Rescale values to specified range.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rmvCSVprefix.html">rmvCSVprefix()</a></code> </p>
        </td>
        <td><p>Remove "ï.." prefix that is appended to csv header</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scoreGBM.html">scoreGBM()</a></code> </p>
        </td>
        <td><p>Annotate glioblastoma (GBM) subtype  based on Neftel 2019 scoring pipeline.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="snip.html">snip()</a></code> </p>
        </td>
        <td><p>Winsorize values at lower and upper quantiles.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sim2adj.html">sim2adj()</a></code> </p>
        </td>
        <td><p>Compute adjaceny matrix from similary (correlation) matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sparse2dense.html">sparse2dense()</a></code> </p>
        </td>
        <td><p>Convert sparse matrix to dense matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sparse2df.html">sparse2df()</a></code> </p>
        </td>
        <td><p>Convert sparse matrix to data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ulength.html">ulength()</a></code> </p>
        </td>
        <td><p>Number of unique values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="value2col.html">value2col()</a></code> </p>
        </td>
        <td><p>Convert values to color gradient</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wideDF2namedList.html">wideDF2namedList()</a></code> </p>
        </td>
        <td><p>Convert wide data.frame to named list</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-seurat-functions" class="hasAnchor"><a href="#section-seurat-functions" class="anchor"></a>Seurat Functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="balanceSamples.html">balanceSamples()</a></code> </p>
        </td>
        <td><p>Subsample cells in seurat object to be balanced (sample-size-wise) across conditions.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downsampleSeurat.html">downsampleSeurat()</a></code> </p>
        </td>
        <td><p>Downsample single cell data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mergeSeuratList.html">mergeSeuratList()</a></code> </p>
        </td>
        <td><p>Merge list of seurat objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="miko_integrate.html">miko_integrate()</a></code> </p>
        </td>
        <td><p>scRNAseq integration wrapper</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="neighborPurity.html">neighborPurity()</a></code> </p>
        </td>
        <td><p>Compute purity of each cell's neighborhood, as defined by KNN graph.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepSeurat.html">prepSeurat()</a></code> </p>
        </td>
        <td><p>prep Seurat</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepSeurat2.html">prepSeurat2()</a></code> </p>
        </td>
        <td><p>prep Seurat (Extended adaptation of prepSeurat)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="recodeBarcode.html">recodeBarcode()</a></code> </p>
        </td>
        <td><p>Recode (i.e., relabel) metadata in Seurat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rmDuplicateGenes.html">rmDuplicateGenes()</a></code> </p>
        </td>
        <td><p>Remove duplicate genes from Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scNormScale.html">scNormScale()</a></code> </p>
        </td>
        <td><p>Normalize and Scale scRNAseq Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetDimRed.html">subsetDimRed()</a></code> </p>
        </td>
        <td><p>Get dimensional reduction from Seurat Object for subset of data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="uniqueFeatures.html">uniqueFeatures()</a></code> </p>
        </td>
        <td><p>Get unique features from metadata column in seurat object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="updateDimNames.html">updateDimNames()</a></code> </p>
        </td>
        <td><p>Ensure that all dimNames are correctly specified in Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wnn_Components.html">wnn_Components()</a></code> </p>
        </td>
        <td><p>Compute network component UMAPs and visualize component weights.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wnn_Run.html">wnn_Run()</a></code> </p>
        </td>
        <td><p>Run WNN Multi-Modal Integration.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-gene-representation" class="hasAnchor"><a href="#section-gene-representation" class="anchor"></a>Gene Representation</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="checkGeneRep.html">checkGeneRep()</a></code> </p>
        </td>
        <td><p>Check gene representation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectSpecies.html">detectSpecies()</a></code> </p>
        </td>
        <td><p>Determine species based on gene representation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ens2sym.so.html">ens2sym.so()</a></code> </p>
        </td>
        <td><p>Convert gene ensemble to symbol in seurat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sym2ens.html">sym2ens()</a></code> <code><a href="sym2ens.html">ensembl2sym()</a></code> </p>
        </td>
        <td><p>Convert gene symbol to ensembl</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="entrez2sym.html">entrez2sym()</a></code> </p>
        </td>
        <td><p>Convert entrez id to gene symbol</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="firstup.html">firstup()</a></code> </p>
        </td>
        <td><p>Uppercase first letter and lowercase rest</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="inferSpecies.html">inferSpecies()</a></code> </p>
        </td>
        <td><p>Infer species from list of Ensemble ids</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepGeneList.html">prepGeneList()</a></code> </p>
        </td>
        <td><p>Prepare gene to ensemble conversion vector</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="speciesConvert.html">speciesConvert()</a></code> </p>
        </td>
        <td><p>Convert gene symbol representation to Hs or Mm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sym2entrez.html">sym2entrez()</a></code> </p>
        </td>
        <td><p>Convert gene symbol to entrez</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-dashboard-utilities" class="hasAnchor"><a href="#section-dashboard-utilities" class="anchor"></a>Dashboard Utilities</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="addLogEntry.html">addLogEntry()</a></code> </p>
        </td>
        <td><p>Add entry to analysis log</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="barcodeLabels.html">barcodeLabels()</a></code> </p>
        </td>
        <td><p>Subset and assign labels to seurat</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flex.asDT.html">flex.asDT()</a></code> </p>
        </td>
        <td><p>Outputs datatable with print button options</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flex.multiTabLogs.html">flex.multiTabLogs()</a></code> </p>
        </td>
        <td><p>Generate multi-tab analysis log list for flexdashboard</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flex.multiTabPlot.html">flex.multiTabPlot()</a></code> </p>
        </td>
        <td><p>Generate multi-tab ggplot handle list for flexdashboard</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flex.multiTabPlotly.html">flex.multiTabPlotly()</a></code> </p>
        </td>
        <td><p>Generate multi-tab list of plotly figures for flexdashboard</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flex.multiTabTables.html">flex.multiTabTables()</a></code> </p>
        </td>
        <td><p>Generate multi-tab datatable list for flexdashboard</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLoadPath.html">getLoadPath()</a></code> </p>
        </td>
        <td><p>Return load path</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="initiateLog.html">initiateLog()</a></code> </p>
        </td>
        <td><p>Initiate analysis log</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadCellRanger.html">loadCellRanger()</a></code> </p>
        </td>
        <td><p>Load CellRanger preprocessed data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadMat.html">loadMat()</a></code> </p>
        </td>
        <td><p>Load gene x cell count matrix into seurat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadMoffat.html">loadMoffat()</a></code> </p>
        </td>
        <td><p>Load preprocessed data from Moffat lab sciRNA-seq3 pipeline</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="updateCentralLog.html">updateCentralLog()</a></code> </p>
        </td>
        <td><p>Update central log</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-save-functions" class="hasAnchor"><a href="#section-save-functions" class="anchor"></a>Save Functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="saveHTML.html">saveHTML()</a></code> </p>
        </td>
        <td><p>Save figure as html</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="savePDF.html">savePDF()</a></code> </p>
        </td>
        <td><p>Save figure as pdf</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://scholar.google.ca/scholar?hl=en&as_sdt=0%2C5&q=mikolajewicz+nicholas&oq='>Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


