<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference • scMiko</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMiko</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="addLogEntry.html">addLogEntry()</a></code> </p>
        </td>
        <td><p>Add entry to analysis log</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="avgGroupExpression.html">avgGroupExpression()</a></code> </p>
        </td>
        <td><p>Get summary of group expression in Seurat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="baderPathways.html">`{`()</a></code> </p>
        </td>
        <td><p>Pathway annotations from Bader Lab</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="balanceMatrixSize.html">balanceMatrixSize()</a></code> </p>
        </td>
        <td><p>Balance matrix dimensions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkGeneRep.html">checkGeneRep()</a></code> </p>
        </td>
        <td><p>Check gene representation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cleanFilterGenes.html">cleanFilterGenes()</a></code> </p>
        </td>
        <td><p>Clean and filter gene list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cluster.UMAP.html">cluster.UMAP()</a></code> </p>
        </td>
        <td><p>UMAP stratified by cluster ID</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dist2hclust.html">dist2hclust()</a></code> </p>
        </td>
        <td><p>Hierarchially-cluster distance matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downsampleSeurat.html">downsampleSeurat()</a></code> </p>
        </td>
        <td><p>Downsample single cell data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="enrichGO.fisher.html">enrichGO.fisher()</a></code> </p>
        </td>
        <td><p>Over-representation enrichment of GO terms using weighted fisher method</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ens2sym.so.html">ens2sym.so()</a></code> </p>
        </td>
        <td><p>Convert gene ensemble to symbol in seurat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="entrez2sym.html">entrez2sym()</a></code> </p>
        </td>
        <td><p>Convert entrez id to gene symbol</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="expression.Plot.html">expression.Plot()</a></code> </p>
        </td>
        <td><p>Violin plot of single cell gene expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="firstup.html">firstup()</a></code> </p>
        </td>
        <td><p>Uppercase first letter and lowercase rest</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fixBarcodeLabel.html">fixBarcodeLabel()</a></code> </p>
        </td>
        <td><p>Fix barcode labels</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flex.asDT.html">flex.asDT()</a></code> </p>
        </td>
        <td><p>Outputs datatable with print buttom options</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flex.multiTabLogs.html">flex.multiTabLogs()</a></code> </p>
        </td>
        <td><p>Generate multi-tab analysis log list for flexdashboard</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flex.multiTabPlot.html">flex.multiTabPlot()</a></code> </p>
        </td>
        <td><p>Generate multi-tab ggplot handle list for flexdashboard</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flex.multiTabPlotly.html">flex.multiTabPlotly()</a></code> </p>
        </td>
        <td><p>Generate multi-tab list of plotly figures for flexdashboard</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flex.multiTabTables.html">flex.multiTabTables()</a></code> </p>
        </td>
        <td><p>Generate multi-tab datatable list for flexdashboard</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generateHeat.html">generateHeat()</a></code> </p>
        </td>
        <td><p>Generate heatmap with plotly interface</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneRepCurve.html">geneRepCurve()</a></code> </p>
        </td>
        <td><p>Plot relationship showing percentage of cells expressing atleast percentage of genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneSets.html">geneSets</a></code> </p>
        </td>
        <td><p>List of scMiko gene sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geom_split_violin.html">geom_split_violin()</a></code> </p>
        </td>
        <td><p>Split violin plot using ggplot2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAnnotationPathways.html">getAnnotationPathways()</a></code> </p>
        </td>
        <td><p>Returns list of annotations for given Entrez gene IDs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAvailableFiles.html">getAvailableFiles()</a></code> </p>
        </td>
        <td><p>Get list of available files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getConnectivity.html">getConnectivity()</a></code> </p>
        </td>
        <td><p>Gene connectivity within network.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getExpressingCells.html">getExpressingCells()</a></code> </p>
        </td>
        <td><p>Get cells that express query gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getExpressionMatrix.html">getExpressionMatrix()</a></code> </p>
        </td>
        <td><p>Get expression matrix from Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getHeat.html">getHeat()</a></code> </p>
        </td>
        <td><p>Create heatmap object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLoadPath.html">getLoadPath()</a></code> </p>
        </td>
        <td><p>Return load path</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getModuleGenes.html">getModuleGenes()</a></code> </p>
        </td>
        <td><p>Get list of module genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getModulePreservation.html">getModulePreservation()</a></code> </p>
        </td>
        <td><p>Determine module preservation between reference and query network</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getNodesEdges.html">getNodesEdges()</a></code> </p>
        </td>
        <td><p>Get nodes and edges from igraph data.frame for visNetwork</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getOrderedGroups.html">getOrderedGroups()</a></code> </p>
        </td>
        <td><p>Get vector of unique ordered group names from Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSoftThreshold.html">getSoftThreshold()</a></code> </p>
        </td>
        <td><p>Analysis of scale free topology for soft-threshold</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSoftThreshold.v2.html">getSoftThreshold.v2()</a></code> </p>
        </td>
        <td><p>Analysis of scale free topology for soft-threshold. Modified from getSoftThreshold.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="id2geneset.html">id2geneset()</a></code> </p>
        </td>
        <td><p>Get Reactome/GO geneset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="id2term.html">id2term()</a></code> </p>
        </td>
        <td><p>Map Reactome/GO ID to term</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="inititateLog.html">inititateLog()</a></code> </p>
        </td>
        <td><p>Initiate analysis log</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="isGeneAvailable.html">isGeneAvailable()</a></code> </p>
        </td>
        <td><p>Check if gene is avaialble in Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m1.analysisParameters.html">m1.analysisParameters()</a></code> <code><a href="m1.analysisParameters.html">m1.analysisParameters()</a></code> </p>
        </td>
        <td><p>Load parameter specifications for Module 1</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m1.barcodeLabels.html">m1.barcodeLabels()</a></code> </p>
        </td>
        <td><p>Assign barcodes labels to seurat metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m1.filterSeurat.html">m1.filterSeurat()</a></code> </p>
        </td>
        <td><p>Apply QC filters to Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m1.getMitoContent.html">m1.getMitoContent()</a></code> </p>
        </td>
        <td><p>Calculate mitochondrial content</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m1.inferSpecies.html">m1.inferSpecies()</a></code> </p>
        </td>
        <td><p>Infer species from list of Ensemble ids</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m1.loadCellRanger.html">m1.loadCellRanger()</a></code> </p>
        </td>
        <td><p>Load CellRanger preprocessed data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m1.loadMoffat.html">m1.loadMoffat()</a></code> </p>
        </td>
        <td><p>Load preprocessed data from Moffat lab sciRNA-seq3 pipeline</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m1.loadTPM.html">m1.loadTPM()</a></code> </p>
        </td>
        <td><p>Load preprocessed TPM data (e.g., Neftel 2019 datasets)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m1.scNormScale.html">m1.scNormScale()</a></code> </p>
        </td>
        <td><p>Normalize and Scale Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m23.binarizeExpression.html">m23.binarizeExpression()</a></code> <code><a href="m23.binarizeExpression.html">m23.binarizeExpression()</a></code> </p>
        </td>
        <td><p>Stratify gene expression into high and low expression groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m9.importMarkers.html">m9.importMarkers()</a></code> </p>
        </td>
        <td><p>M9. Import Gene List</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapSubsetSeurat.html">mapSubsetSeurat()</a></code> </p>
        </td>
        <td><p>Filter seurat object according to specified meta data entries</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mergeSeuratList.html">mergeSeuratList()</a></code> </p>
        </td>
        <td><p>Merge list of seurat objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="modulePackages.html">modulePackages()</a></code> </p>
        </td>
        <td><p>Load Module-Specific Packages</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multiLevel.FindMarkers.html">multiLevel.FindMarkers()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="namedList2longDF.html">namedList2longDF()</a></code> </p>
        </td>
        <td><p>Convert named list to long data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="namedList2wideDF.html">namedList2wideDF()</a></code> </p>
        </td>
        <td><p>Convert named list to wide data.frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="networkProperties.Plot.html">networkProperties.Plot()</a></code> </p>
        </td>
        <td><p>Plot network properties for different soft thresholds</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="optimalDS.html">optimalDS()</a></code> </p>
        </td>
        <td><p>Cut clustered tree at varying heights and overlay with dendrogram to find optimal parameter set.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepExpression.html">prepExpression()</a></code> </p>
        </td>
        <td><p>Normalize and scale data within appropriate assay</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepGeneList.html">prepGeneList()</a></code> </p>
        </td>
        <td><p>prep Gene List</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepSeurat.html">prepSeurat()</a></code> </p>
        </td>
        <td><p>prep Seurat</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="QC.scatterPlot.html">QC.scatterPlot()</a></code> </p>
        </td>
        <td><p>QC scatter plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="QC.violinPlot.html">QC.violinPlot()</a></code> </p>
        </td>
        <td><p>QC violin plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qNorm.html">qNorm()</a></code> </p>
        </td>
        <td><p>Quantile Normalization of 2 Vectors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="recaleValues.html">BaCo()</a></code> </p>
        </td>
        <td><p>Bayesian Correlation algorithm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rescaleValues.html">rescaleValues()</a></code> </p>
        </td>
        <td><p>Rescale values to specified range.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rmDuplicateGenes.html">rmDuplicateGenes()</a></code> </p>
        </td>
        <td><p>Remove duplicate genes from Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rmvCSVprefix.html">rmvCSVprefix()</a></code> </p>
        </td>
        <td><p>Remove "ï.." prefix that is appended to csv header</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runEnrichment.html">runEnrichment()</a></code> </p>
        </td>
        <td><p>Wrapper to run  GO enrichment, using fisher enrichment method</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runWGCNA.html">runWGCNA()</a></code> </p>
        </td>
        <td><p>run WGCNA analysis on scRNAseq expression matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="savePlot.html">savePlot()</a></code> </p>
        </td>
        <td><p>Save ggplot as image file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaleTOM.html">scaleTOM()</a></code> </p>
        </td>
        <td><p>TOM matrix rescaling prior to computing consensus topological overlap.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scExpression.UMAP.html">scExpression.UMAP()</a></code> </p>
        </td>
        <td><p>Cell-level gene expression projected on UMAP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scMikoReload.html">scMikoReload()</a></code> </p>
        </td>
        <td><p>reload scMiko</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="searchAnnotations.html">searchAnnotations()</a></code> </p>
        </td>
        <td><p>Search Reactome/GO databases for terms that match query</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="setResolution.html">setResolution()</a></code> </p>
        </td>
        <td><p>Set cluster resolution</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sim2adj.html">sim2adj()</a></code> </p>
        </td>
        <td><p>Compute adjaceny matrix from similary (correlation) matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="speciesConvert.html">speciesConvert()</a></code> </p>
        </td>
        <td><p>Convert gene symbol representation to Hs or Mm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetSeurat.html">subsetSeurat()</a></code> </p>
        </td>
        <td><p>Subset Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sym2ens.html">sym2ens()</a></code> </p>
        </td>
        <td><p>Convert gene symbol to ensembl</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sym2entrez.html">sym2entrez()</a></code> </p>
        </td>
        <td><p>Convert gene symbol to entrez</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="term2id.html">term2id()</a></code> </p>
        </td>
        <td><p>Map Reactome/GO term to ID</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="updateGeneSets.html">updateGeneSets()</a></code> </p>
        </td>
        <td><p>Create/update list of gene sets for scMiko package.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="value2col.html">value2col()</a></code> </p>
        </td>
        <td><p>Convert values to color gradient</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="variableGenes.Plot.html">variableGenes.Plot()</a></code> </p>
        </td>
        <td><p>Plot variable genes in Seurat Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wgcna2graphDF.html">wgcna2graphDF()</a></code> </p>
        </td>
        <td><p>Convert adjacency or TOM matrix to igraph data.frame</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://scholar.google.ca/scholar?hl=en&as_sdt=0%2C5&q=mikolajewicz+nicholas&oq='>Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


